src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs

```diff
@@ -9,94 +9,106 @@
+            public sealed override Task WriteAsync(ReadOnlyMemory<byte> destination, CancellationToken cancellationToken) => throw new NotSupportedException();
 
-        public sealed override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
-        {
-            ValidateBufferArgs(buffer, offset, count);
-            return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
-        }
+            public sealed override int Read(byte[] buffer, int offset, int count)
+            {
+                ValidateBufferArgs(buffer, offset, count);
+                return ReadAsync(new Memory<byte>(buffer, offset, count), CancellationToken.None).GetAwaiter().GetResult();
+            }
 
-        public sealed override void CopyTo(Stream destination, int bufferSize) =>
-            CopyToAsync(destination, bufferSize, CancellationToken.None).GetAwaiter().GetResult();
+            public sealed override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                ValidateBufferArgs(buffer, offset, count);
+                return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
+            }
 
-        public virtual bool NeedsDrain => false;
+            public sealed override void CopyTo(Stream destination, int bufferSize) =>
+                CopyToAsync(destination, bufferSize, CancellationToken.None).GetAwaiter().GetResult();
 
-        public virtual Task<bool> DrainAsync(int maxDrainBytes)
-        {
-            Debug.Assert(false, "DrainAsync should not be called for this response stream");
-            return Task.FromResult(false);
-        }
+            public virtual bool NeedsDrain => false;
 
-        protected override void Dispose(bool disposing)
-        {
-            if (disposing)
+            public virtual Task<bool> DrainAsync(int maxDrainBytes)
             {
-                if (NeedsDrain)
+                Debug.Fail($"DrainAsync should not be called for this response stream: {GetType()}");
+                return Task.FromResult(false);
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                // Only attempt draining if we haven't started draining due to disposal; otherwise
+                // multiple calls to Dispose (which happens frequently when someone disposes of the
+                // response stream and response content) will kick off multiple concurrent draining
+                // operations. Also don't delegate to the base if Dispose has already been called,
+                // as doing so will end up disposing of the connection before we're done draining.
+                if (_disposed)
+                {
+                    return;
+                }
+                _disposed = true;
```

Comment:
Should the check/set be atomic for thread safety?

Reply:
Yeah, probably.